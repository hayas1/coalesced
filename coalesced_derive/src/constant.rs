use syn::{parse_quote, Ident, Path};

pub const DERIVE_CONSTRUCTION: &str = "Construction";

#[derive(Debug, Clone)]
pub struct Constant {
    pub path_semigroup: Path,
    pub _path_annotated_semigroup: Path,
    pub ident_semigroup_op: Ident,
    pub path_annotated: Path,
    pub path_reversed: Path,
}
pub trait ConstantExt {
    fn constant() -> Constant;
}
pub enum Absolute {}
impl ConstantExt for Absolute {
    fn constant() -> Constant {
        Constant {
            path_semigroup: parse_quote!(::coalesced::Semigroup),
            _path_annotated_semigroup: parse_quote!(::coalesced::AnnotatedSemigroup),
            ident_semigroup_op: parse_quote!(semigroup_op),
            path_annotated: parse_quote!(::coalesced::Annotated),
            path_reversed: parse_quote!(::coalesced::op::reverse::Reversed),
        }
    }
}
#[cfg(feature = "use_scope")]
pub enum Use {}
#[cfg(feature = "use_scope")]
impl ConstantExt for Use {
    fn constant() -> Constant {
        Constant {
            path_semigroup: parse_quote!(Semigroup),
            _path_annotated_semigroup: parse_quote!(AnnotatedSemigroup),
            ident_semigroup_op: parse_quote!(semigroup_op),
            path_annotated: parse_quote!(Annotated),
            path_reversed: parse_quote!(Reversed),
        }
    }
}
