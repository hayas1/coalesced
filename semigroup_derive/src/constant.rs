use syn::{parse_quote, Attribute, Path, TypeParam};

pub const DERIVE_CONSTRUCTION: &str = "Construction";
pub const DERIVE_SEMIGROUP: &str = "Semigroup";

#[derive(Debug, Clone)]
pub struct Constant {
    pub path_semigroup: Path,
    pub path_annotated_semigroup: Path,
    pub path_annotated: Path,
    pub path_annotate: Path,
    pub path_monoid: Path,
    pub path_commutative: Path,
    pub path_construction_trait: Path,
    pub path_construction_annotated: Path,
    pub path_construction_monoid: Path,
    pub default_type_param: TypeParam,
    pub attr_feature_monoid: Option<Attribute>,
    pub attr_feature_async: Option<Attribute>,
}
pub trait ConstantExt {
    fn constant() -> Constant;
}
pub enum External {}
impl ConstantExt for External {
    fn constant() -> Constant {
        Constant {
            path_semigroup: parse_quote! {::semigroup::Semigroup},
            path_annotated_semigroup: parse_quote! {::semigroup::AnnotatedSemigroup},
            path_annotated: parse_quote! {::semigroup::Annotated},
            path_annotate: parse_quote! {::semigroup::Annotate},
            path_monoid: parse_quote! {::semigroup::Monoid},
            path_commutative: parse_quote! {::semigroup::Commutative},
            path_construction_trait: parse_quote! {::semigroup::Construction},
            path_construction_annotated: parse_quote! {::semigroup::ConstructionAnnotated},
            path_construction_monoid: parse_quote! {::semigroup::ConstructionMonoid},
            default_type_param: parse_quote! { A },
            attr_feature_monoid: None,
            attr_feature_async: None,
        }
    }
}
pub enum Internal {}
impl ConstantExt for Internal {
    fn constant() -> Constant {
        Constant {
            path_semigroup: parse_quote! {crate::Semigroup},
            path_annotated_semigroup: parse_quote! {crate::AnnotatedSemigroup},
            path_annotated: parse_quote! {crate::Annotated},
            path_annotate: parse_quote! {crate::Annotate},
            path_monoid: parse_quote! {crate::Monoid},
            path_commutative: parse_quote! {crate::Commutative},
            path_construction_trait: parse_quote! {crate::Construction},
            path_construction_annotated: parse_quote! {crate::ConstructionAnnotated},
            path_construction_monoid: parse_quote! {crate::ConstructionMonoid},
            default_type_param: parse_quote! { A },
            attr_feature_monoid: Some(parse_quote! {#[cfg(feature = "monoid")]}),
            attr_feature_async: Some(parse_quote! {#[cfg(feature = "async")]}),
        }
    }
}
