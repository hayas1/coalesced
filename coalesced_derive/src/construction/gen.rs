use proc_macro2::TokenStream;
use quote::ToTokens;
use syn::DeriveInput;

use crate::{
    constant::ConstantExt,
    construction::{attr::ConstructionAttr, implementor::Construction},
};

pub fn gen_construction<C: ConstantExt>(derive: &DeriveInput) -> TokenStream {
    let constant = C::constant();
    let attr = match ConstructionAttr::new(derive) {
        Ok(attr) => attr,
        Err(e) => return e.into_compile_error(),
    };
    let construction = match Construction::new(&constant, derive, &attr) {
        Ok(construction) => construction,
        Err(e) => return e.into_compile_error(),
    };
    construction.into_token_stream()
}

#[cfg(test)]
mod tests {
    use crate::constant::{Absolute, Use};

    use super::*;

    #[test]
    fn test_derive_construction_annotation() {
        let derive = syn::parse_quote! {
            #[derive(Construction)]
            #[construction(annotated, op = Coalesce)]
            pub struct Coalesced<T>(pub Option<T>);
        };
        let generated = gen_construction::<Absolute>(&derive);
        let formatted = prettyplease::unparse(&syn::parse2(generated).unwrap());
        insta::with_settings!({ snapshot_path => "../../tests/snapshots" }, {
            insta::assert_snapshot!("annotated", formatted);
        });
    }

    #[test]
    fn test_derive_construction_no_annotateion() {
        let derive = syn::parse_quote! {
            #[derive(ConstructionUse)]
            #[construction(op = Coalesce)]
            pub struct Coalesced<T>(pub Option<T>);
        };
        let generated = gen_construction::<Use>(&derive);
        let formatted = prettyplease::unparse(&syn::parse2(generated).unwrap());
        insta::with_settings!({ snapshot_path => "../../tests/snapshots" }, {
            insta::assert_snapshot!("not_annotated", formatted);
        });
    }
}
