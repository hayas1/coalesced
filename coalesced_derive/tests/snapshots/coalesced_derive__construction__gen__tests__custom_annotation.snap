---
source: coalesced_derive/src/construction/gen.rs
assertion_line: 67
expression: formatted
---
pub trait Coalesce: Sized + Semigroup {
    fn coalesce(self, other: Self) -> Self {
        ::coalesced::Semigroup::semigroup_op(self, other)
    }
}
impl<T: IntoIterator + FromIterator<T::Item>> Coalesce for Concatenated<T> {}
impl<T: IntoIterator + FromIterator<T::Item>> Coalesce
for ::coalesced::op::reverse::Reversed<Concatenated<T>> {}
impl<
    T: IntoIterator + FromIterator<T::Item>,
    X: IntoIterator + FromIterator<X::Item>,
> Coalesce for ::coalesced::Annotated<Concatenated<T>, X> {}
impl<
    T: IntoIterator + FromIterator<T::Item>,
    X: IntoIterator + FromIterator<X::Item>,
> Coalesce
for ::coalesced::op::reverse::Reversed<::coalesced::Annotated<Concatenated<T>, X>> {}
impl<T: IntoIterator + FromIterator<T::Item>> ::coalesced::Semigroup
for Concatenated<T> {
    fn semigroup_op(base: Self, other: Self) -> Self {
        Self::default_semigroup_op(base, other, vec![(); 0], vec![(); 0])
    }
}
impl<T: IntoIterator + FromIterator<T::Item>> From<T> for Concatenated<T> {
    fn from(value: T) -> Self {
        Concatenated(value)
    }
}
impl<T: IntoIterator + FromIterator<T::Item>> Concatenated<T> {
    pub fn into_inner(self) -> T {
        self.0
    }
}
impl<T: IntoIterator + FromIterator<T::Item>> std::ops::Deref for Concatenated<T> {
    type Target = T;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}
impl<T: IntoIterator + FromIterator<T::Item>> std::ops::DerefMut for Concatenated<T> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}
